blog/_posts/2018-06-04-scalac-profiling.md

```diff
@@ -0,0 +1,1328 @@
+
+* How to use `scalac-profiling` to replicate a similar analysis.
+* Why Shapeless-based code is prone to slow compilation times if not used with
+  care.
+* How you can replicate a similar analysis of compilation times on code that
+  abuses implicit searches and macros.
+* How implicit search and macros interact in unexpected ways that hurt
+  productivity and how you can optimize their interaction.
+
+The most important take-away from this guide is that *you should not take
+slow Scala compile times for granted*. It's worth investigating why slow
+compiles happens as we can often reduce them on our side!
+
+### Pointers to read the article
+
+If you want to apply the same technique in your Scala projects or you have no
+previous knowledge of the reasons why automatic typeclass derivation is slow,
+reading the whole blog post is **highly recommended**.
+
+If you're already familiar with the source of inefficiencies or you don't
+have much time, jump directly to the
+[detective work][#the-quest-for-optimization] that digs into
+the profiling data.
+
+## TOC
+
+- [TOC](#toc)
+- [The codebase](#the-codebase)
+- [The setup and workflow](#the-setup-and-workflow)
+  - [Compiling the codebase](#compiling-the-codebase)
+    - [Warm up the compiler](#warm-up-the-compiler)
+- [The profiling toolkit and theory](#the-profiling-toolkit-and-theory)
+  - [Compiler statistics](#compiler-statistics)
+    - [Setting statistics up](#setting-statistics-up)
+  - [Walking into the lion's den](#walking-into-the-lions-den)
+  - [The troublemaker](#the-troublemaker)
+  - [An initial exploration of the data](#an-initial-exploration-of-the-data)
+    - [A profiling plugin for `scalac`](#a-profiling-plugin-for-scalac)
+    - [The first visual](#the-first-visual)
+    - [Typeclass derivation for the win](#typeclass-derivation-for-the-win)
+    - [The cost of implicit macros](#the-cost-of-implicit-macros)
+    - [The world of shapeless](#the-world-of-shapeless)
+    - [Quick derivation example](#quick-derivation-example)
+- [The quest for optimization](#the-quest-for-optimization)
+    - [Reading the implicit search flamegraph](#reading-the-implicit-search-flamegraph)
+    - [A tour through Shapeless's `Generic`](#a-tour-through-shapelesss-generic)
+    - [The Strict/Lazy macro doesn't like the aux pattern](#the-strict-lazy-macro-doesnt-like-the-aux-pattern)
+    - [Deduplicating more expansions](#deduplicating-more-expansions)
+    - [Getting the final results](#getting-the-final-results)
+- [Conclusion](#conclusion)
+
+This is a long blog post. Put on your profiling hat and let's get our hands
+dirty!
+
+## The codebase
+
+[Bloop](https://github.com/scalacenter/bloop) is a *build-tool-agnostic*
+compilation server with a focus on developer productivity that I developed at
+the Scala Center together with [Martin Duhem](https://github.com/Duhemm). It
+gives you about ~20-25% faster compilation times than sbt, and we plan on
```

Comment:
"20-25% faster compilation times than sbt" makes bloop sound like a drop-in replacement for sbt.